Bottom: cfdc14642e6934ed67efeb1bc2611ff2ee669245
Top:    8db877bf1edc750dd665b8dab73418c3dcf62c05
Author: Peter Clifton <peter@clifton-electronics.co.uk>
Date:   2015-01-04 20:07:07 +0000

Add hysteresis to grid-snapping


---

diff --git a/src/crosshair.c b/src/crosshair.c
index 78b8111..6ae12e9 100644
--- a/src/crosshair.c
+++ b/src/crosshair.c
@@ -927,6 +927,11 @@ FitCrosshairIntoGrid (Coord X, Coord Y)
   struct snap_data snap_data;
   int ans;
 
+  Coord old_x, old_y;
+
+  old_x = Crosshair.X;
+  old_y = Crosshair.Y;
+
   Crosshair.X = CLAMP (X, Crosshair.MinX, Crosshair.MaxX);
   Crosshair.Y = CLAMP (Y, Crosshair.MinY, Crosshair.MaxY);
 
@@ -953,11 +958,26 @@ FitCrosshairIntoGrid (Coord X, Coord Y)
     }
 
   snap_data.crosshair = &Crosshair;
-  snap_data.nearest_sq_dist =
-    crosshair_sq_dist (&Crosshair, nearest_grid_x, nearest_grid_y);
   snap_data.nearest_is_grid = true;
   snap_data.x = nearest_grid_x;
   snap_data.y = nearest_grid_y;
+#if 0
+  snap_data.nearest_sq_dist = crosshair_sq_dist (&Crosshair, snap_data.x, snap_data.y);
+
+  if (snap_data.nearest_sq_dist > PCB->Grid / 3 * PCB->Grid / 3)
+    {
+      snap_data.x = old_x;
+      snap_data.y = old_y;
+    }
+#endif
+
+  if (labs (nearest_grid_x - Crosshair.X) > PCB->Grid / 3)
+    snap_data.x = old_x;
+
+  if (labs (nearest_grid_y - Crosshair.Y) > PCB->Grid / 3)
+    snap_data.y = old_y;
+
+  snap_data.nearest_sq_dist = crosshair_sq_dist (&Crosshair, snap_data.x, snap_data.y);
 
   ans = NO_TYPE;
   if (!PCB->RatDraw)
